lxd/vm_qemu.go

```diff
@@ -0,0 +1,2329 @@
+		}
+
+		d, err := device.New(vm, vm.state, "", newDevice, nil, nil)
+		if err != nil {
+			return []string{} // Couldn't create Device, so this cannot be an update.
+		}
+
+		_, updateFields := d.CanHotPlug()
+		return updateFields
+	})
+
+	// Do some validation of the config diff.
+	err = containerValidConfig(vm.state.OS, vm.expandedConfig, false, true)
+	if err != nil {
+		return errors.Wrap(err, "Invalid expanded config")
+	}
+
+	// Do full expanded validation of the devices diff.
+	err = containerValidDevices(vm.state, vm.state.Cluster, vm.Name(), vm.expandedDevices, true)
+	if err != nil {
+		return errors.Wrap(err, "Invalid expanded devices")
+	}
+
+	// Use the device interface to apply update changes.
+	err = vm.updateDevices(removeDevices, addDevices, updateDevices, oldExpandedDevices)
+	if err != nil {
+		return err
+	}
+
+	// Update MAAS (must run after the MAC addresses have been generated).
+	updateMAAS := false
+	for _, key := range []string{"maas.subnet.ipv4", "maas.subnet.ipv6", "ipv4.address", "ipv6.address"} {
+		if shared.StringInSlice(key, updateDiff) {
+			updateMAAS = true
+			break
+		}
+	}
+
+	if !vm.IsSnapshot() && updateMAAS {
+		err = vm.maasUpdate(oldExpandedDevices.CloneNative())
+		if err != nil {
+			return err
+		}
+	}
+
+	// Finally, apply the changes to the database
+	err = query.Retry(func() error {
+		tx, err := vm.state.Cluster.Begin()
+		if err != nil {
+			return err
+		}
+
+		// Snapshots should update only their descriptions and expiry date.
+		if vm.IsSnapshot() {
+			err = db.InstanceSnapshotUpdate(tx, vm.id, vm.description, vm.expiryDate)
+			if err != nil {
+				tx.Rollback()
+				return errors.Wrap(err, "Snapshot update")
+			}
+		} else {
+			err = db.ContainerConfigClear(tx, vm.id)
+			if err != nil {
+				tx.Rollback()
+				return err
+
+			}
+			err = db.ContainerConfigInsert(tx, vm.id, vm.localConfig)
+			if err != nil {
+				tx.Rollback()
+				return errors.Wrap(err, "Config insert")
+			}
+
+			err = db.ContainerProfilesInsert(tx, vm.id, vm.project, vm.profiles)
+			if err != nil {
+				tx.Rollback()
+				return errors.Wrap(err, "Profiles insert")
+			}
+
+			err = db.DevicesAdd(tx, "instance", int64(vm.id), vm.localDevices)
+			if err != nil {
+				tx.Rollback()
+				return errors.Wrap(err, "Device add")
+			}
+
+			err = db.ContainerUpdate(tx, vm.id, vm.description, vm.architecture, vm.ephemeral, vm.expiryDate)
+			if err != nil {
+				tx.Rollback()
+				return errors.Wrap(err, "Container update")
+			}
+
+		}
+
+		if err := db.TxCommit(tx); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "Failed to update database")
+	}
+
+	err = writeBackupFile(vm)
+	if err != nil && !os.IsNotExist(err) {
+		return errors.Wrap(err, "Failed to write backup file")
+	}
+
+	// Success, update the closure to mark that the changes should be kept.
+	undoChanges = false
+
+	var endpoint string
+
+	// tomp TODO container references in events.
+	if vm.IsSnapshot() {
```

Comment:
@stgraber also, these events, should they be changed as per the other event question earlier?

Reply:
yep, should be `virtual-machine-updated` and we'll want the paths to point to `1/.0/virtual-machines` for consistency